### PR TITLE
Multi-z

### PR DESCRIPTION
**Pull request by Kurfursten**
**Please Note:** This Multi-Z module will not affect maps that are not configured to be multi-Z. I have _no intention_ presently of pursuing a new map project, this is merely to give us the option in the future. I assume this would be a natural extension for a theoretical snow map, if one were to exist.

Ports Multi-Z from Polaris.

Includes
* Movable atoms fall between levels, and fall damage
* Multi-Z pipes
* Ability to look down through open space and see the area below you
* Original multiz test map. Even smaller than test_tiny, this has 4 levels. Level 1 is empty and inaccessible. Level 2 is a subterranean mine accessible by ladder and stairs. Level 3 is a snowy "surface" level. Level 4 is the sky level, accessible from stairs in the research lab.
* Coders: Some notes on port documentation in multiz\_HELPERS.dm; debug verb update_all_open_spaces() available under debug tab

Known Issues:
* Open Spaces not being regularly updated by OSC
* Z-shadow'd mobs (mobs seen from above on lower levels) do not update their direction to match their real counterpart
* Z-shadow'd mobs do not seem to be relaying speech properly (? Hard to verify)
* Non-mobs also not relaying speech between z-level open spaces (e.g.: speaker, vending machine)
* Upward/downward pipes sprite missing
* Upward/downward pipes need to be tuned to work with our pipe layering (currently try to attach to any properly-facing pipe regardless of layer)
* Objects (structures, items) on lower levels are not properly tinted like turfs are to indicate they are on a lower floor.
* Stairs do not like directional density (e.g.: placing a directional window on the top of stairs will prevent traveling down them even if it would not normally block passage)
* Building lattices/catwalks on open space does not work
* Throwing items not working perfectly (hacky solution was to make them fall immediately instead of finishing their arc)

This PR does not yet implement (but may, with support)
* MultiZ power cables
* Elevators
* Traveling upward (jetpack, magboot scaling)
* MultiZ ZAS (e.g.: creating a fire below an open space will not cause heat to rise into the upper level)

Example image from test_multiz
untitled

### CODE REVIEW

**Comment by ghost** [comment]
Y E S

**Comment by Durgen** [comment]
>Level 3 is snowy 

003c07697e54384d0f1b2edb7daeb7d197fbf3652e71cef81b1437806ddd6f74

**Comment by Kammerjunk** [comment]
Hey this is pretty cool

**Comment by SonixApache** [comment]
absolute madman

**Comment by SysadminJeroen** [comment]
Holy shit.

Op 23 jan. 2018 08:13 schreef "SonixApache" <notifications@github.com>:

File: code/modules/mob/mob.dm [hunk 1]
```diff
@@ -2095,5 +2095,8 @@ mob/proc/on_foot()
 				alpha = initial(alpha)
 				alphas.Remove(source_define)
 
+/mob/proc/is_dead()
```
Comment (DamianX) [inline code/modules/mob/mob.dm:2098]: see: `/mob/proc/isDead()`

File: code/modules/multiz/basic.dm [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+/proc/get_zstep(ref, dir)
+	if(dir == UP)
+		. = GetAbove(ref)
+	else if (dir == DOWN)
+		. = GetBelow(ref)
+	else
+		. = get_step(ref, dir)
+
+/atom/proc/CanPass(atom/movable/mover, turf/target, height=1.5, air_group = 0)
```
Comment (DamianX) [inline code/modules/multiz/basic.dm:55]: we use `Cross` for this

File: code/modules/multiz/movement.dm [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+	playsound(loc, "punch", 25, 1, -1)
+	var/damage = 15 // Because wounds heal rather quickly, 15 should be enough to discourage jumping off but not be enough to ruin you, at least for the first time.
+	apply_damage(rand(0, damage), BRUTE, LIMB_HEAD)
+	apply_damage(rand(0, damage), BRUTE, LIMB_CHEST)
+	apply_damage(rand(0, damage), BRUTE, LIMB_LEFT_LEG)
+	apply_damage(rand(0, damage), BRUTE, LIMB_RIGHT_LEG)
+	apply_damage(rand(0, damage), BRUTE, LIMB_LEFT_ARM)
+	apply_damage(rand(0, damage), BRUTE, LIMB_RIGHT_ARM)
+	AdjustKnockdown(4)
+	updatehealth()
```
Comment (DamianX) [inline code/modules/multiz/movement.dm:310]: `apply_damage` should do this already

File: vgstation13.dme [hunk 4]
```diff
@@ -2261,7 +2271,7 @@
 #include "interface\web\interface.dms"
 #include "maprendering\maprendering.dm"
 #include "maps\_map.dm"
-#include "maps\tgstation.dm"
+#include "maps\test_multiz.dm"
```
Comment (DamianX) [inline vgstation13.dme:2274]: don't forget to change this

File: __DEFINES/process.dm [hunk 5]
```diff
@@ -0,0 +1,15 @@
+// Process status defines
+#define PROCESS_STATUS_IDLE 1
+#define PROCESS_STATUS_QUEUED 2
+#define PROCESS_STATUS_RUNNING 3
+#define PROCESS_STATUS_MAYBE_HUNG 4
+#define PROCESS_STATUS_PROBABLY_HUNG 5
+#define PROCESS_STATUS_HUNG 6
+
+// Process time thresholds
+#define PROCESS_DEFAULT_HANG_WARNING_TIME 300 // 30 seconds
```
Comment (DamianX) [inline __DEFINES/process.dm:10]: use `30 SECONDS` and such where applicable

File: __DEFINES/subsystem.dm [hunk 6]
```diff
@@ -55,3 +57,4 @@
 #define SS_DISPLAY_TICKER         -10
 #define SS_DISPLAY_UNSPECIFIED     0
 #define SS_DISPLAY_SUN             10
+#define SS_DISPLAY_OSC			   20
```
Comment (DamianX) [inline __DEFINES/subsystem.dm:60]: the other two are `OS` while this is `OSC`, this is upsetting

File: code/modules/multiz/_HELPERS.dm [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	var/turf/edge_type ///< What the map edge should be formed with. (null = world.turf)
+
+/turf/proc/is_space()
+	return 0
+
+/turf/space/is_space()
+	return 1
+
+/atom/proc/set_dir(new_dir) //We should probably use this for our whole codebase for shadows. If z-shadow isn't working I may have to implement it widely.
+	. = new_dir != dir
```
Comment (DamianX) [inline code/modules/multiz/_HELPERS.dm:26]: This return value is never used as far as I can tell, and we already have `change_dir`. Is this supposed to act differently?

**Review by DamianX — CHANGES_REQUESTED** [review]

**Comment by DamianX** [comment]
also we /pipebursting/ now?

**Comment by SonixApache** [comment]
multi-z pipematter cascades
the end times are upon us
i cannot wait

**Comment by Lokiamis** [comment]
Cool, dog

**Comment by RemieRichards** [comment]
You really shouldn't port an old dusty mutli z right now, Byond 512 made multiz SUCH an easier feature to implement, with instantaneous updates of things moving around on lower zs without (server) lag.

I'd recommend you look into replacing the OSC with 512's ```visual_contents``` variable.

**Comment by Exxion** [comment]
Technically multiZAS is already in the code, enabled with a define
But there's basically no chance it actually works correctly

**Comment by PJB3005** [comment]
@RemieRichards we still can't use 512 until viewport freeze is fixed.

**Comment by Praecordia** [comment]
NANI?

**Comment by Blithering** [comment]
does this mean we can have clown traitors killing people by dropping pianos down to a lower z level on their heads?

**Comment by ShiftyRail** [comment]
@RemieRichards @PJB3005 It's also that 512 won't even compile on our codebase.
#17094 
http://www.byond.com/forum/?post=2338657

**Comment by Kurfursten** [comment]
For the record, this is officially on hold until we get 512 compatibility as I have heard it's coming "soon"

If I'm gonna be writing this I may as well make it visual_contents compatible and embrace the future.

**Comment by ghost** [comment]
@Kurfursten fair warning: 512 is ***extremely***   unstable, tg had to blacklist a few versions because they were so buggy, they were basically wallhacks

**Comment by Kurfursten** [comment]
are we 512 now? I heard we might be

**Comment by Kammerjunk** [comment]
Pomf said we are in the last thread in [this post](http://boards.4chan.org/vg/thread/208119468#p208303125)
image

**Comment by DamianX** [comment]
maybe one day

**Comment by Anton-Kr** [comment]
is this merged yet

File: code/modules/multiz/basic.dm (hunk 2 above)
Comment (ghost) [reply code/modules/multiz/basic.dm:55]: Fixed in #30178

File: code/modules/mob/mob.dm (hunk 1 above)
Comment (ghost) [reply code/modules/mob/mob.dm:2098]: Since removed

File: code/modules/multiz/movement.dm (hunk 3 above)
Comment (ghost) [reply code/modules/multiz/movement.dm:310]: Fixed in #30178

File: __DEFINES/process.dm (hunk 5 above)
Comment (kane-f) [reply __DEFINES/process.dm:10]: Fixed in #30178

File: __DEFINES/subsystem.dm (hunk 6 above)
Comment (ghost) [reply __DEFINES/subsystem.dm:60]: OSC since removed

File: __DEFINES/process.dm (hunk 5 above)
Comment (ghost) [reply __DEFINES/process.dm:10]: File since removed, was unused

**Pull request closed without merging**